### PR TITLE
ARROW-12751: [C++] Implement minimum/maximum kernels

### PR DESCRIPTION
**Pull request by lidavidm**
This is a bit messy, but implements a variadic scalar maximum/minimum kernel.

### CODE REVIEW

File: cpp/src/arrow/compute/api_scalar.h [hunk 1]
```diff
@@ -42,6 +42,11 @@ struct ArithmeticOptions : public FunctionOptions {
   bool check_overflow;
 };
 
+struct ARROW_EXPORT MinMaxOptions : public FunctionOptions {
```
Comment (lidavidm) [inline cpp/src/arrow/compute/api_scalar.h:45]: Do we want to name this something else?

File: cpp/src/arrow/compute/api_scalar.h [hunk 2]
```diff
@@ -250,6 +255,24 @@ Result<Datum> Power(const Datum& left, const Datum& right,
                     ArithmeticOptions options = ArithmeticOptions(),
                     ExecContext* ctx = NULLPTR);
 
+/// \brief
```
Comment (lidavidm) [inline cpp/src/arrow/compute/api_scalar.h:258]: I forgot to fill this in, will fix.

File: cpp/src/arrow/compute/kernels/scalar_arithmetic.cc [hunk 3]
```diff
@@ -428,6 +454,45 @@ ArrayKernelExec ArithmeticExecFromOp(detail::GetTypeId get_id) {
   }
 }
 
+template <template <typename... Args> class KernelGenerator, typename Op>
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:457]: This is rather a lot of code being generated; it might be nice if I could figure out the right template setup to consolidate the temporal types with their respective equivalent integral type.

File: cpp/src/arrow/ipc/json_simple.cc [hunk 4]
```diff
@@ -911,6 +912,26 @@ Status DictArrayFromJSON(const std::shared_ptr<DataType>& type,
       .Value(out);
 }
 
+Status ScalarFromJSON(const std::shared_ptr<DataType>& type,
```
Comment (lidavidm) [inline cpp/src/arrow/ipc/json_simple.cc:915]: I'm going to port this into ARROW-12859.

File: cpp/src/arrow/testing/gtest_util.cc [hunk 5]
```diff
@@ -360,6 +360,30 @@ void AssertDatumsEqual(const Datum& expected, const Datum& actual, bool verbose)
       break;
   }
 }
+ARROW_TESTING_EXPORT void AssertDatumsApproxEqual(const Datum& expected,
```
Comment (lidavidm) [inline cpp/src/arrow/testing/gtest_util.cc:363]: I'm going to port this to ARROW-12859 (and fix the unimplemented case).

File: cpp/src/arrow/ipc/json_simple.h [hunk 6]
```diff
@@ -30,6 +30,7 @@ namespace arrow {
 
 class Array;
 class DataType;
+class Scalar;
```
Comment (lidavidm) [inline cpp/src/arrow/ipc/json_simple.h:33]: This is the wrong declaration (should be `struct Scalar`).

File: cpp/src/arrow/compute/kernel.h [hunk 7]
```diff
@@ -52,7 +52,7 @@ struct ARROW_EXPORT KernelState {
 /// \brief Context/state for the execution of a particular kernel.
 class ARROW_EXPORT KernelContext {
  public:
-  explicit KernelContext(ExecContext* exec_ctx) : exec_ctx_(exec_ctx) {}
+  explicit KernelContext(ExecContext* exec_ctx) : exec_ctx_(exec_ctx), state_() {}
```
Comment (cyb70289) [inline cpp/src/arrow/compute/kernel.h:55]: This change is already in current code. Needs rebase?

File: cpp/src/arrow/compute/api_scalar.h (hunk 1 above)
Comment (ianmcook) [reply cpp/src/arrow/compute/api_scalar.h:45]: I think it is worth giving this a more general name, because there are other potential uses. Maybe `ElementWiseAggregateOptions`? (since this is aggregation, just across instead of down)

Comment (ianmcook) [reply cpp/src/arrow/compute/api_scalar.h:45]: If using "Aggregate" is too confusing, maybe "Combine" or "Merge" would work too. e.g. `ElementWiseCombineOptions`

Comment (lidavidm) [reply cpp/src/arrow/compute/api_scalar.h]: Since we have `ScalarAggregateOptions` I think `ElementWiseAggregateOptions` should be ok. I've updated the PR.

File: cpp/src/arrow/compute/kernels/scalar_arithmetic.cc (hunk 3 above)
Comment (bkietz) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:457]: There's GeneratePhysicalInteger, though that doesn't include floating point. Perhaps you could add GeneratePhysicalNumeric?

File: cpp/src/arrow/compute/kernels/scalar_arithmetic.cc [hunk 8]
```diff
@@ -453,6 +518,26 @@ struct ArithmeticFunction : ScalarFunction {
...
+    RETURN_NOT_OK(CheckArity(*values));
+
+    using arrow::compute::detail::DispatchExactImpl;
+    if (auto kernel = DispatchExactImpl(this, *values)) return kernel;
+
+    EnsureDictionaryDecoded(values);
+
+    if (auto type = CommonNumeric(*values)) {
+      ReplaceTypes(type, values);
+    }
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:534]: (Perhaps as a follow-up:)
```suggestion
    } else if (auto type = CommonTimestamp(*values)) {
      ReplaceTypes(type, values);
    } else if (auto type = CommonBinary(*values)) {
      ReplaceTypes(type, values);
    }
```

File: cpp/src/arrow/compute/kernels/scalar_arithmetic.cc [hunk 9]
```diff
@@ -516,6 +601,203 @@ std::shared_ptr<ScalarFunction> MakeUnarySignedArithmeticFunctionNotNull(
...
+    ArrayData* output = out->mutable_array();
+
+    // Exactly one array (output = input)
+    if (batch.values.size() == 1) {
+      *output = *batch[0].array();
+      return Status::OK();
+    }
+
+    // At least one array, two or more arguments
+    int64_t length = 0;
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:656]: This should already be present in ExecBatch::length

File: cpp/src/arrow/compute/kernels/scalar_arithmetic.cc [hunk 10]
```diff
@@ -516,6 +601,203 @@ std::shared_ptr<ScalarFunction> MakeUnarySignedArithmeticFunctionNotNull(
...
+      }
+      // This won't work for nested or variable-sized types
+      std::memcpy(output->buffers[1]->mutable_data(),
+                  input.buffers[1]->data() + (input.offset * sizeof(OutValue)),
+                  length * sizeof(OutValue));
+    }
+
+    for (size_t i = 1; i < batch.values.size(); i++) {
+      OutputArrayWriter<OutType> writer(out->mutable_array());
+      if (batch.values[i].is_scalar()) {
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:730]: I think instead it'd be useful to sort the inputs and handle all the scalars first. This would allow you to consider only arrays in this loop

**Review by bkietz — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/compute/kernels/scalar_arithmetic.cc (hunk 3 above)
Comment (lidavidm) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic.cc]: I added this. Note that if we want to have `CommonTimestamp()` I don't think this'll fly since we'd need to scale the timestamp values to the common unit before comparison.

Comment (bkietz) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic.cc]: No, CommonTimestamp indicates that an implicit cast is necessary (which in the case of timestamps includes the scaling). It's not the responsibility of the kernel to execute that implicit cast

Comment (lidavidm) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic.cc]: Ah! Ok, my mistake. I'll add some test cases for that.

Binary types aren't implemented here, though (and would need some work/restructuring to handle).

File: cpp/src/arrow/compute/kernels/codegen_internal.h [hunk 11]
```diff
@@ -1093,6 +1097,41 @@ ArrayKernelExec GeneratePhysicalInteger(detail::GetTypeId get_id) {
...
+    case Type::UINT16:
+      return Generator<UInt16Type, Args...>::Exec;
+    case Type::UINT32:
+      return Generator<UInt32Type, Args...>::Exec;
+    case Type::UINT64:
+      return Generator<UInt64Type, Args...>::Exec;
+    case Type::FLOAT:
+      return Generator<FloatType, Args...>::Exec;
+    case Type::DOUBLE:
+      return Generator<DoubleType, Args...>::Exec;
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/codegen_internal.h:1128]:
```suggestion
    case Type::FLOAT:
      return Generator<FloatType, Args...>::Exec;
    case Type::DOUBLE:
      return Generator<DoubleType, Args...>::Exec;
    default:
      return GeneratePhysicalInteger<Generator, Args...>(get_id);
```

File: cpp/src/arrow/compute/kernels/scalar_arithmetic.cc [hunk 12]
```diff
@@ -516,6 +562,202 @@ std::shared_ptr<ScalarFunction> MakeUnarySignedArithmeticFunctionNotNull(
...
+    if (batch.values.size() == 1) {
+      *output = *batch[0].array();
+      return Status::OK();
+    }
+
+    // At least one array, two or more arguments
+    DCHECK_GE(first_array_index, 0);
+    DCHECK_LT(first_array_index, batch.values.size());
+    DCHECK(batch.values[first_array_index].is_array());
+    if (any_scalar) {
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:631]: Handling scalars should happen before the output == input bail. If the folded scalar is null then we can either ignore it (skip_nulls=true) or emit a null array without examining any array inputs (skip_nulls=false). After that we're guaranteed to have a valid scalar (for which MakeArrayFromScalar will not materialize a validity bitmap, so you can drop the BitmapXor etc)

File: cpp/src/arrow/compute/kernels/scalar_arithmetic.cc [hunk 13]
```diff
@@ -516,6 +562,202 @@ std::shared_ptr<ScalarFunction> MakeUnarySignedArithmeticFunctionNotNull(
...
+                  batch.length * sizeof(OutValue));
+    }
+
+    if (!options.skip_nulls) {
+      // We can precompute the validity buffer in this case
+      // AND together the validity buffers of all arrays
+      ARROW_ASSIGN_OR_RAISE(output->buffers[0], ctx->AllocateBitmap(batch.length));
+      bool first = true;
+      for (const auto& arg : batch.values) {
+        if (!arg.is_array()) continue;
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:680]: instead of dealing with datums, you could build an ArrayDataVector and push the non scalar inputs (and the promoted folded scalar) into that. It'd make it explicit that you're only dealing with arrays and be a bit more readable, I think

File: cpp/src/arrow/compute/kernels/codegen_internal.h (hunk 11 above)
Comment (lidavidm) [reply cpp/src/arrow/compute/kernels/codegen_internal.h:1128]: Unfortunately the two generators are slightly different since it's `GeneratePhysicalInteger<Generator, Type0, Args>` which will stamp out `Generator<Type0, Int8Type, Args>...` vs here where it's `GeneratePhysicalNumeric<Generator, Args>`.

**Comment by lidavidm** [comment]
(N.B. I still need to fix CommonTimestamp, got caught up in fixing the other feedback)

File: cpp/src/arrow/compute/kernels/codegen_internal.h (hunk 11 above)
Comment (pitrou) [reply cpp/src/arrow/compute/kernels/codegen_internal.h:1128]: Perhaps, but the discrepancy seems a bit gratuitous.

File: cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc [hunk 14]
```diff
@@ -1161,5 +1255,242 @@ TYPED_TEST(TestUnaryArithmeticFloating, AbsoluteValue) {
...
+  this->Assert(Maximum, this->scalar("0"), {this->scalar("NaN"), this->scalar("0")});
+  this->Assert(Maximum, this->scalar("Inf"), {this->scalar("Inf"), this->scalar("NaN")});
+  this->Assert(Maximum, this->scalar("Inf"), {this->scalar("NaN"), this->scalar("Inf")});
+  this->Assert(Maximum, this->scalar("-Inf"),
+               {this->scalar("-Inf"), this->scalar("NaN")});
+  this->Assert(Maximum, this->scalar("-Inf"),
+               {this->scalar("NaN"), this->scalar("-Inf")});
+  this->Assert(Maximum, this->scalar("NaN"), {this->scalar("NaN"), this->scalar("null")});
+  this->Assert(Minimum, this->scalar("0"), {this->scalar("0"), this->scalar("Inf")});
+  this->Assert(Minimum, this->scalar("-Inf"), {this->scalar("0"), this->scalar("-Inf")});
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc:1339]: Can you add some test with `-0.0` and `0.0`?

**Review by pitrou — CHANGES_REQUESTED** [review]
Can you explain what the point is of a variadic function when we don't have e.g. variadic addition?

**Comment by ianmcook** [comment]
> Can you explain what the point is of a variadic function when we don't have e.g. variadic addition?

My understanding is that we are aiming for Arrow's compute API to reach parity with the collections of built-in functions available in popular SQL engines. These SQL engines include several variadic functions that combine/merge/aggregate values row-wise. Two such SQL functions are `greatest()` and `least()`. The two kernels implemented in this PR are equivalent to those two SQL functions. Another such SQL function is `concat()` and its variant `concat_ws()`, which we intend to implement in ARROW-12709.

There are two ways we could implement this functionality without implementing variadic functions:
1. Chain multiple calls to a binary function (as is required for addition of more than two values, as you mention)
2. Combine the arrays row-wise into a ListArray, then operate on the ListArray with a unary function

If one of those alternative approaches is superior, then perhaps we do not need a variadic function. But my current understanding is that both of these alternative approaches are inferior for reasons of usability and efficiency respectively.

**Comment by pitrou** [comment]
I see. Well, if efficiency is really important and it's common to have more than 2 inputs, then why not (though the implementation in this PR is probably far from optimal).

File: cpp/src/arrow/compute/kernels/scalar_arithmetic.cc [hunk 15]
```diff
@@ -516,6 +565,190 @@ std::shared_ptr<ScalarFunction> MakeUnarySignedArithmeticFunctionNotNull(
...
+
+    // Exactly one array to consider (output = input)
+    if (arrays.size() == 1) {
+      *output = *arrays[0];
+      return Status::OK();
+    }
+
+    // Two or more arrays to consider
+    if (scalar_count > 0) {
+      // We allocated the last array from a scalar: recycle it as the output
```
[the remark this answers is not in the document]
Comment (lidavidm) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:649]: Good catch, thanks.

File: cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc (hunk 14 above)
Comment (lidavidm) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc:1328]: Done.

File: docs/source/cpp/compute.rst [hunk 16]
```diff
@@ -310,6 +310,20 @@ output element is null.
...
 
+These functions take any number of inputs of numeric type (in which case they
+will be cast to the :ref:`common numeric type <common-numeric-type>` before
+comparison) or of temporal types. If any input is dictionary encoded it will be
+expanded for the purposes of comparison.
+
++--------------------------+------------+---------------------------------------------+---------------------+---------------------------------------+-------+
+| Function names           | Arity      | Input types                                 | Output type         | Options class                         | Notes |
++==========================+============+=============================================+=====================+=======================================+=======+
+| maximum, minimum         | Varargs    | Numeric and Temporal                        | Numeric or Temporal | :struct:`ElementWiseAggregateOptions` | \(1)  |
```
Comment (pitrou) [inline docs/source/cpp/compute.rst:321]: It just came to my mind that it's gonna be very confusing to have "minimum" and "maximum" functions in addition to "min_max". Perhaps we can be less ambiguous, e.g. "scalar_min", "scalar_max" (or "horizontal_min" or "row_min"...).

Comment (lidavidm) [reply docs/source/cpp/compute.rst:321]: That's a good point. I'll rename the compute function. Perhaps `elementwise_min` in accordance with the options struct (though that is a bit wordy)?

Comment (pitrou) [reply docs/source/cpp/compute.rst:321]: That's wordy indeed. Or at least `element_wise_min`.

Comment (ianmcook) [reply docs/source/cpp/compute.rst:321]: How about naming them `least` and `greatest` like they do in SQL?

Comment (pitrou) [reply docs/source/cpp/compute.rst:321]: How is that less ambiguous?

Comment (lidavidm) [reply docs/source/cpp/compute.rst]: I just changed it to element_wise_min/max.

File: cpp/src/arrow/compute/kernels/scalar_arithmetic.cc [hunk 17]
```diff
@@ -721,7 +720,7 @@ struct ScalarMinMax {
         }
       }
     }
-    output->null_count = -1;
+    output->null_count = output->buffers[0] ? -1 : 0;
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:723]:
```suggestion
    output->null_count = output->buffers[0] ? kUnknownNullCount : 0;
```

File: cpp/src/arrow/compute/kernels/scalar_arithmetic.cc [hunk 18]
```diff
@@ -516,6 +564,190 @@ std::shared_ptr<ScalarFunction> MakeUnarySignedArithmeticFunctionNotNull(
...
+    const auto descrs = batch.GetDescriptors();
+    const size_t scalar_count =
+        static_cast<size_t>(std::count_if(batch.values.begin(), batch.values.end(),
+                                          [](const Datum& d) { return d.is_scalar(); }));
+    if (scalar_count == batch.values.size()) {
+      ExecScalar(batch, options, out->scalar().get());
+      return Status::OK();
+    }
+
+    ArrayData* output = out->mutable_array();
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:612]: Instead of trying to recycle buffers, we *could* change the kernel's flag and allow the executor to preallocate the output data buffer. Then we populate this with the anti extreme (for maximum/uint64_t we zero the buffer). Then we wouldn't need to use a bitmap to express "no values folded yet"; `get_extreme({anti_extreme, v...}) == get_extreme({v...})`. We'd also be able to compute the null bitmap separately in either case: `!skip_nulls` AND the bitmaps, `skip_nulls` OR the bitmaps.

In the case where `scalar_count != 0`, we can ExecScalar and use the result instead of an anti_extreme

Comment (lidavidm) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:612]: This works quite well except in the case that we have a singular NaN array argument (`element_wise_min([NaN])`) - in which case we want to emit `NaN` not the antiextreme.

Comment (lidavidm) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:612]: Ah, maybe let's just have the 'antiextreme' of a float be NaN.

Comment (bkietz) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:642]: The antiextremes of float are Inf, -Inf

Comment (lidavidm) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:642]: Inf would trip up if we had something like `element_wise_min([NaN, 0], [NaN, 1])` in which case we'd want to emit `NaN` and not `Inf` as the first element. Using `Inf` as the antiextreme means we'd emit `Inf` since the kernel takes any non-NaN value over any NaN.

**Comment by ianmcook** [comment]
@lidavidm could you briefly describe what happens here if the arrays passed to these kernels have different but comparable, data types such as one array of doubles and one array of integers?

**Comment by lidavidm** [comment]
> @lidavidm could you briefly describe what happens here if the arrays passed to these kernels have different but comparable data types such as one array of doubles and one array of integers?

For numeric/temporal types, all arrays will be cast to a common compatible type, much the same as with arithmetic kernels (e.g. for doubles + integers, all values will be cast to doubles).

File: cpp/src/arrow/compute/kernels/scalar_arithmetic.cc (hunk 18 above)
Comment (bkietz) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:642]: Huh, I was expecting `fmin(NaN, x) = NaN` but in fact it's `fmin(NaN, x) = x`. Nevermind

**Review by bkietz — APPROVED** [review]
Thanks for doing this!
